### DISCUSSION TITLE
Using a SHA for remote actions not working as expected

### DISCUSSION THREAD

**Discussion by riker09** [community/community#26660] [Actions]
<p>I’m using a SHA as a reference for a remote action in one of my workflows, basically like this:</p>
<pre><code class="lang-auto">    - name: Deploy functions
      uses: w9jds/firebase-action@0cd4655ce2f0fb91fa1ac6454df8fde67772d9ea
      with:
        args: deploy --only functions,firestore --force
      env:
        FIREBASE_TOKEN: ${{ secrets.FIREBASE_TOKEN }}
        PROJECT_ID: ${{ needs.prepare.outputs.pid }}
</code></pre>
<p>The SHA points to a specific commit in that repository. This repo has upgraded their dependecies recently which resultet in a failing deploy action in my workflow. Upon investigating the issue I discovered that the actual <code>docker run</code> command is <em>ALWAYS</em> using the <code>:latest</code> tag.</p>
<p>I tried using a tagged version like <code>:v.2.0.0</code> but that used the same <code>:latest</code> tag.</p>
<p>Here’s the output of the failed step:</p>
<pre><code class="lang-auto">/usr/bin/docker run --name w9jdsfirebaseactionlatest_723405 --label 08450d --workdir /github/workspace --rm -e FIREBASE_PROJECT_ID -e FIREBASE_TOKEN -e PROJECT_ID -e INPUT_ARGS -e HOME -e GITHUB_JOB -e GITHUB_REF -e GITHUB_SHA -e GITHUB_REPOSITORY -e GITHUB_REPOSITORY_OWNER -e GITHUB_RUN_ID -e GITHUB_RUN_NUMBER -e GITHUB_RETENTION_DAYS -e GITHUB_RUN_ATTEMPT -e GITHUB_ACTOR -e GITHUB_WORKFLOW -e GITHUB_HEAD_REF -e GITHUB_BASE_REF -e GITHUB_EVENT_NAME -e GITHUB_SERVER_URL -e GITHUB_API_URL -e GITHUB_GRAPHQL_URL -e GITHUB_REF_NAME -e GITHUB_REF_PROTECTED -e GITHUB_REF_TYPE -e GITHUB_WORKSPACE -e GITHUB_ACTION -e GITHUB_EVENT_PATH -e GITHUB_ACTION_REPOSITORY -e GITHUB_ACTION_REF -e GITHUB_PATH -e GITHUB_ENV -e GITHUB_STEP_SUMMARY -e RUNNER_OS -e RUNNER_ARCH -e RUNNER_NAME -e RUNNER_TOOL_CACHE -e RUNNER_TEMP -e RUNNER_WORKSPACE -e ACTIONS_RUNTIME_URL -e ACTIONS_RUNTIME_TOKEN -e ACTIONS_CACHE_URL -e GITHUB_ACTIONS=true -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/_temp/_runner_file_commands":"/github/file_commands" -v "/home/runner/work/my/project":"/github/workspace" w9jds/firebase-action:latest deploy --only functions,firestore --force
</code></pre>
<p>(pay attention to the end of that command output where it is using <code>w9jds/firebase-action:latest</code>)</p>
<p>How is the SHA supposed to work? And shouldn’t it rather throw an error if the SHA cannot be found instead of silently using the <code>:latest</code> tag?</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>Here’s the reason:</p><aside class="onebox githubblob" data-onebox-src="https://github.com/w9jds/firebase-action/blob/0cd4655ce2f0fb91fa1ac6454df8fde67772d9ea/action.yaml#L9">
  <header class="source">

      <a href="https://github.com/w9jds/firebase-action/blob/0cd4655ce2f0fb91fa1ac6454df8fde67772d9ea/action.yaml#L9" target="_blank" rel="noopener">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/w9jds/firebase-action/blob/0cd4655ce2f0fb91fa1ac6454df8fde67772d9ea/action.yaml#L9" target="_blank" rel="noopener">w9jds/firebase-action/blob/0cd4655ce2f0fb91fa1ac6454df8fde67772d9ea/action.yaml#L9</a></h4>



    <pre class="onebox">      <code class="lang-yaml">
        <ol class="start lines" start="1" style="counter-reset: li-counter 0 ;">
            <li>name: 'GitHub Action for Firebase'</li>
            <li>author: 'Jeremy Shore'</li>
            <li>description: 'Wraps the firebase-tools CLI to enable common commands.'</li>
            <li>branding:</li>
            <li>  icon: 'package'</li>
            <li>  color: 'gray-dark'</li>
            <li>runs:</li>
            <li>  using: 'docker'</li>
            <li class="selected">  image: 'docker://w9jds/firebase-action:latest'</li>
        </ol>
      </code>
    </pre>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>You get the <em>action</em> you’re referencing at exactly that commit, but unfortunately that action references a container image it’s using with <code>latest</code>.</p>
<p>So the answer is: The <code>uses</code> parameter is working as it should, but unfortunately that action isn’t well made (at least not in the sense of running reproducibly).</p>

**Comment by riker09** [comment]
<p>Thank you for your fast reply and the excellent explanation. Guess that is one more thing one has to keep in mind when reviewing the code of an action.</p>
